### ISSUE TITLE
Holding non-layer data from plugins in the viewer (or some other store)

### ISSUE THREAD

**Issue by jni** [napari/napari#2985]
## 🚀 Feature

Sometimes plugins need to store things that aren't represented by layers, such as dataframes, floats, networkx graphs, or really any random object. Currently their only option is to create a heavyweight object that contains all the variables they need. It would be nice if napari provided a namespace for plugins to store non-layer data.

## Pitch

One implementation idea would be a simple nested dictionary keyed by plugin import name, then variable names. The same thing could also be implemented as nested namespaces, e.g. `viewer.vars.zarpaint.padded_data`.

## Alternatives

- plugins can keep managing their own objects. The downside is that you can't have any cross-plugin talk.
- A global namespace (external even to the viewer) for all variables. The downside could be name clashes. And global things come with their own problems, e.g. do we really want separate viewers to share these variables...
- something else...?

**Comment by alisterburt** [comment]
I really like this idea - if it's a choice between the nested dictionary and the nested namespaces I think I prefer the nested namespaces.

If we wanted to really encourage plugin crosstalk we could expose this 'non-layer' data as an [`EventedModel`](https://github.com/napari/napari/blob/804cebee7c0f60fcec1513f5908638c8330deaa2/napari/utils/events/evented_model.py)
We could imagine allowing plugin devs to provide a simple dict and [create the model dynamically](https://pydantic-docs.helpmanual.io/usage/models/#dynamic-model-creation) for them... possible new 'provide data' hook spec?

**Comment by goanpeca** [comment]
Would there be a need to also "protect" some of these fields?

In Python anything can be accessed but would some of these fields need some extra "protection" to avoid accidental tampering from other plugins?

**Comment by alisterburt** [comment]
good point! If we did want to use the `EventedModel` there is some support for [faux immutability](https://pydantic-docs.helpmanual.io/usage/models/#faux-immutability)
edit: I'm not sure if the immutability can easily be switched on and off on an existing model instance
